core/src/main/java/io/questdb/cairo/sql/SqlExecutionCircuitBreakerWrapper.java

```diff
@@ -0,0 +1,139 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.sql;
+
+import io.questdb.std.MemoryTag;
+import io.questdb.std.Misc;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.TestOnly;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+// does not have to be thread-safe, because worker threads own the jobs/contexts
+// which hold SqlExecutionCircuitBreakerWrapper objects
+public class SqlExecutionCircuitBreakerWrapper implements SqlExecutionCircuitBreaker, Closeable {
+    private SqlExecutionCircuitBreaker delegate;
+    private NetworkSqlExecutionCircuitBreaker networkSqlExecutionCircuitBreaker;
+
+    @Override
+    public void cancel() {
+        delegate.cancel();
+    }
+
+    @Override
+    public boolean checkIfTripped(long millis, long fd) {
+        return delegate.checkIfTripped(millis, fd);
+    }
+
+    @Override
+    public boolean checkIfTripped() {
+        return delegate.checkIfTripped();
+    }
+
+    @Override
+    public void close() throws IOException {
+        delegate = Misc.freeIfCloseable(networkSqlExecutionCircuitBreaker);
+    }
+
+    @Override
+    public @Nullable SqlExecutionCircuitBreakerConfiguration getConfiguration() {
+        return delegate.getConfiguration();
+    }
+
+    @TestOnly
+    public SqlExecutionCircuitBreaker getDelegate() {
+        return delegate;
+    }
+
+    @Override
+    public long getFd() {
+        return delegate.getFd();
+    }
+
+    @Override
+    public int getState() {
+        return delegate.getState();
+    }
+
+    @Override
+    public int getState(long millis, long fd) {
+        return delegate.getState(millis, fd);
+    }
+
+    public void init(SqlExecutionCircuitBreakerWrapper wrapper) {
+        init(wrapper.delegate);
+    }
+
+    public void init(SqlExecutionCircuitBreaker executionContextCircuitBreaker) {
+        final SqlExecutionCircuitBreakerConfiguration sqlExecutionCircuitBreakerConfiguration = executionContextCircuitBreaker.getConfiguration();
+        if (sqlExecutionCircuitBreakerConfiguration != null) {
+            if (networkSqlExecutionCircuitBreaker == null) {
+                networkSqlExecutionCircuitBreaker = new NetworkSqlExecutionCircuitBreaker(sqlExecutionCircuitBreakerConfiguration, MemoryTag.NATIVE_CB2);
+            } else {
+                networkSqlExecutionCircuitBreaker.resetTimer();
+            }
+            delegate = networkSqlExecutionCircuitBreaker;
+        } else {
+            delegate = executionContextCircuitBreaker;
```

Comment:
Here we rely on an implicit convention: when `executionContextCircuitBreaker.getConfiguration()` is `null`, the CB has to be thread-safe as we'll be sharing it between multiple threads. On the other hand, when the configuration is not null, the assumption is that this is a network CB. So, at least it's worth documenting this behavior in javadoc.

As one more step, we could solve that by introducing an `isThreadSafe()` (or `isNetworkCB()`) method. In this case, we could pre-create the `networkSqlExecutionCircuitBreaker` object similarly to how we did that in `PageFrameReduceJob`.

Reply:
I like the `isThreadSafe()` method idea, i do not like that we rely on the null check either.
Originally I started with a `SqlExecutionCircuitBreakerType` enum and the circuit breaker implementations had a `getType()` method, but eventually removed it.
Will add `isThreadSafe()`.

Re pre-creating `networkSqlExecutionCircuitBreaker`, I thought about it but decided to use lazy initialisation, because when we are using AtomicBoolean or Noop circuit breakers, we do not need a NetworkCB instance at all, and we would create an object for nothing.

Reply:
> Re pre-creating `networkSqlExecutionCircuitBreaker`, I thought about it but decided to use lazy initialisation, because when we are using AtomicBoolean or Noop circuit breakers, we do not need a NetworkCB instance at all, and we would create an object for nothing.

Creation of a network CB is kinda inevitable as long as users send queries over HTTP or PGWire and that should stand true pretty much for all instances.

Reply:
that is true, will remove lazy init.

Reply:
added `isThreadSafe()` and removed lazy init of `networkSqlExecutionCircuitBreaker`